mmdet/apis/ote/extension/utils/runner.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+import time
+
+import torch.distributed as dist
+from mmcv.runner import RUNNERS, EpochBasedRunner
+from mmcv.runner import get_dist_info
+
+
+@RUNNERS.register_module()
+class EpochRunnerWithCancel(EpochBasedRunner):
+    """
+    Simple modification to EpochBasedRunner to allow cancelling the training during an epoch. The early stopping hook
+    should set the runner.should_stop flag to True if stopping is required.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(EpochRunnerWithCancel, self).__init__(*args, **kwargs)
+        self.should_stop = False
+        _, world_size = get_dist_info()
+        self.distributed = True if world_size > 1 else False
+
+    def stop(self) -> bool:
+        """ Returning a boolean to break the training loop
+        This method supports distributed training by broadcasting should_stop to other ranks
+        :return: a cancellation bool
+        """
+        if self.rank == 0 and self.should_stop:
+            broadcast_obj = [True]
+        else:
+            broadcast_obj = [None]
+
+        if self.distributed:
+            dist.broadcast_object_list(broadcast_obj, src=0)
```

Comment:
Does this broadcasting have any tangible effect on the training time?

Reply:
I did some benchmark, the time is very marginal, about `0.66` ms